app/src/main/java/com/duckduckgo/app/browser/downloader/DownloaderUtil.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2022 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.downloader
+
+import android.net.Uri
+import android.webkit.MimeTypeMap
+import timber.log.Timber
+import java.lang.IllegalStateException
+import java.util.*
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+object DownloaderUtil {
+
+    // Slightly different pattern than the one used in URLUtil (also contains `inline`).
+    private val CONTENT_DISPOSITION_PATTERN =
+        Pattern.compile(
+            "(inline|attachment)\\s*;\\s*filename\\s*=\\s*(\"((?:\\\\.|[^\"\\\\])*)\"|[^;]*)\\s*",
+            Pattern.CASE_INSENSITIVE,
+        )
+
+    // Generic values for the Content-Type header used to enforce the decision to take the file extension from the URL if possible.
+    private val GENERIC_CONTENT_TYPES = setOf(
+        "application/octet-stream",
+        "application/unknown",
+        "binary/octet-stream",
+    )
+
+    // Inspired from UrlUtil.guessFileName and adapted to fix some scenarios highlighted in tests.
+    fun guessFileName(url: String?, contentDisposition: String?, mimeType: String?): String {
+        var filename: String? = null
+
+        if (contentDisposition != null) {
+            filename = extractFileNameFromContentDisposition(contentDisposition)
+        }
+
+        if (filename.isNullOrEmpty()) {
+            filename = fileNameFromUrl(url)
+        }
+
+        // Dummy name if we couldn't extract it.
+        if (filename.isNullOrEmpty()) {
+            filename = "downloadfile"
+        }
+
+        return filenameWithExtension(filename, mimeType)
+    }
+
+    private fun filenameWithExtension(initialFilename: String, mimeType: String?): String {
+        var filename = initialFilename
+        var extension: String? = null
```

Comment:
this seems to be always null, what's the reason for adding it? you can return the value of extension in the if clause and avoid creating a null String